### PR TITLE
feat: add user stories and tests for new JS certification projects

### PR DESCRIPTION
**Pull request by scissorsneedfoodtoo**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Related to https://github.com/freeCodeCamp/demo-projects/pull/520

This PR converts the existing certification projects for the new JS certification into multi-file projects and adds tests for the new / adapted user stories. It also adds the new Pokémon Search App project, which is new for the project-based JS curriculum.

The only project with a new id is the Pokémon Search App project. The updated projects (Palindrome Checker, Roman Numeral Converter, Telephone Number Validator, and Cash Register), still have the same ids as the current versions of those projects in the current JS curriculum.

### CODE REVIEW

**Review by jdwilkin4 — CHANGES_REQUESTED** [review]
Was able to go through everything 👍

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-palindrome-checker-project/palindrome-checker.md [hunk 1]
```diff
@@ -1,125 +1,381 @@
...
 
 ```js
-assert(palindrome('_eye') === true);
+const el = document.getElementById('result');
+assert(!!el && el.nodeName.toLowerCase() === 'div');
 ```
 
-`palindrome("race car")` should return `true`.
+
+When you click on the `#check-btn` element without entering a value into the `#text-input` element, an alert should appear with the text `Please input a value`.
```
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-palindrome-checker-project/palindrome-checker.md:55]: Thank you for catching this huge issue, @jdwilkin4. I hadn't tested the projects with minimal code, so this is really helpful.

It seems like we can't actually run call `assert` within a global function override like I was trying to do here, so it was as if there was no test at all, so these pass in all cases.

Fortunately it seems like there's a simple fix for this. I'll go in test it with all the projects and double check that they only pass when alerts are triggered with the correct message.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-roman-numeral-converter-project/roman-numeral-converter.md [hunk 2]
```diff
@@ -26,190 +26,424 @@ Convert the given number into a roman numeral.
...
+**Objective:** Build an app that is functionally similar to <a href="https://roman-numeral-converter.freecodecamp.rocks" target="_blank" rel="noopener noreferrer nofollow">https://roman-numeral-converter.freecodecamp.rocks</a>
+
+**User Stories:**
+
+1. You should have an `input` element with an `id` of `number`
+1. You should have a `button` element with an `id` of `convert-btn`
+1. You should have a `div` element with an `id` of `output`
+1. When you click on the `#convert-btn` element without entering a value into the `#number` element, an alert should appear with the text `Please enter a valid number`
+1. When the `#number` element contains the number `-1` and the `#convert-btn` element is clicked, an alert should appear with the text `Please enter a number greater than or equal to 1`
+1. When the `#number` element contains the number `4000` or greater and the `#convert-btn` element is clicked, an alert should appear with the text `Please enter a number less than or equal to 3999`
```
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-roman-numeral-converter-project/roman-numeral-converter.md:38]: Thanks for catching the faulty text and tests for these user stories, @jdwilkin4. I'll update them so we're checking the `innerText` of the `div#output` element instead.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md [hunk 3]
```diff
@@ -1,31 +1,26 @@
...
 dashedName: build-a-cash-register
 ---
 
 # --description--
 
-Design a cash register drawer function `checkCashRegister()` that accepts purchase price as the first argument (`price`), payment as the second argument (`cash`), and cash-in-drawer (`cid`) as the third argument.
+Here you'll build a cash register app that will return change to the customer based on the price of the item, the amount of cash provided by the customer, and the amount of cash in the cash drawer. You'll also need to show different messages to the user in different scenarios, such as when the customer provides too little cash or when the cash drawer doesn't have enough to issue the correct change.
 
-`cid` is a 2D array listing available currency.
+There are a couple of variables that will be provided by the tests for the user stories below:
```
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md:13]: Sure, that sounds good to me.

I'll update the description now and will add those to the JS seed.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-pokemon-search-app-project/build-a-pokemon-search-app.md [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+
+# --description--
+
+In this project, you'll build an app that will search for Pokémon by name or ID and display the results to the user. To retrieve the Pokémon data and images, you'll use freeCodeCamp's <a href="https://pokeapi-proxy.freecodecamp.rocks/" target="_blank" rel="noopener noreferrer nofollow">PokéAPI Proxy</a>.
+
+**Objective:** Build an app that is functionally similar to <a href="https://pokemon-search-app.freecodecamp.rocks" target="_blank" rel="noopener noreferrer nofollow">https://pokemon-search-app.freecodecamp.rocks</a>
+
+**User Stories:**
+
+1. You should have an `input` element with an `id` of `search-input`
```
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-pokemon-search-app-project/build-a-pokemon-search-app.md:17]: This should be fixed now, with a single test to check for an `input` element that's also required.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md [hunk 5]
```diff
@@ -1,252 +1,559 @@
...
+assert(!!el && el.nodeName.toLowerCase() === 'input');
 ```
 
-# --hints--
+You should have a `div` element with an `id` of `change-due`.
 
-`checkCashRegister(19.5, 20, [["PENNY", 1.01], ["NICKEL", 2.05], ["DIME", 3.1], ["QUARTER", 4.25], ["ONE", 90], ["FIVE", 55], ["TEN", 20], ["TWENTY", 60], ["ONE HUNDRED", 100]])` should return an object.
+```js
+const el = document.getElementById('change-due');
+assert(!!el && el.nodeName.toLowerCase() === 'div');
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md:69]:
```suggestion
assert.strictEqual(el?.nodeName?.toLowerCase(), 'div');
```

We generally try to use specific assert methods (easier to debug when tests fail in CI, allows us to use `--fcc-expected--` in hints, etc.)

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
A quick note that applies to all of the assertions.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md (hunk 5 above)
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md]: These should all use Chai assert methods now. Some projects did require multiple assert methods, but if one of those fails the whole block for that test / hint should fail.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md [hunk 6]
```diff
@@ -1,252 +1,559 @@
...
-Design a cash register drawer function `checkCashRegister()` that accepts purchase price as the first argument (`price`), payment as the second argument (`cash`), and cash-in-drawer (`cid`) as the third argument.
+Here you'll build a cash register app that will return change to the customer based on the price of the item, the amount of cash provided by the customer, and the amount of cash in the cash drawer. You'll also need to show different messages to the user in different scenarios, such as when the customer provides too little cash or when the cash drawer doesn't have enough to issue the correct change.
 
-`cid` is a 2D array listing available currency.
+There are a few variables you'll need to use in your code:
 
-The `checkCashRegister()` function should always return an object with a `status` key and a `change` key.
+- `price`: the price of the item as a floating point number
+- `cash`: is amount of cash provided by the customer for the item, which is provided via an `input` element on the page
+- `cid`: cash-in-drawer, a 2D array listing available currency in the cash drawer
```
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md:17]:
```suggestion
- `price`: the price of the item as a floating-point number.
- `cash`: the amount of cash provided by the customer for the item, which is provided via an `input` element on the page.
- `cid`: cash-in-drawer, a 2D array listing available currency in the cash drawer.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md [hunk 7]
```diff
@@ -1,252 +1,559 @@
...
 
-Return `{status: "INSUFFICIENT_FUNDS", change: []}` if cash-in-drawer is less than the change due, or if you cannot return the exact change.
+If you'd like to test your application with different values for `price` and `cid`, make sure to declare them with the `let` keyword so they can be reassigned by our tests.
 
-Return `{status: "CLOSED", change: [...]}` with cash-in-drawer as the value for the key `change` if it is equal to the change due.
+Your application should show different messages depending on price of the item, the amount of cash provided by the customer, and the amount of cash in the drawer:
```
Comment (Sembauke) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md:21]:
```suggestion
Your application should show different messages depending on the price of the item, the amount of cash provided by the customer, and the amount of cash in the drawer:
```

**Review by Sembauke — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md (hunk 7 above)
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md:21]: Thanks for catching this, @Sembauke.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md (hunk 6 above)
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/build-a-cash-register-project/cash-register.md:17]: Thanks for catching these issues, too, @Sembauke.

The one thing I may go back in change is "floating-point number". It seems like we use the unhyphenated version, "floating point number" in other parts of the curriculum, but I can't find the hyphenated version anywhere.

I'll do more digging and possibly change it to "floating point number" for consistency.

**Comment by naomi-lgbt** [comment]
Hi @scissorsneedfoodtoo - looks like a rebase might have exploded on this one. I went ahead and merged `main` back into this PR, so it should be back where you had it. :3

**Review by naomi-lgbt — APPROVED** [review]
Just leaving a note that this should be fine to merge - I didn't see any conflicts or issues with the existing JS cert projects.

**Comment by scissorsneedfoodtoo** [comment]
Thanks @naomi-lgbt, I just saw that. I might have messed something up while rebasing last night.

Hopefully it won't cause issues, or mess with the history too much. But it's no problem to move these changes into a fresh PR, so just let me know if you'd prefer to do that.

**Review by jdwilkin4 — APPROVED** [review]

**Review by Sembauke — APPROVED** [review]

**Pull request merged**